Refresh/Purge API

phg49389 opened the pull request:
I've been writing this on-and-off for the past month or so, when all my other projects were waiting on something. This should resolve Issue #2 if we implement it.

There's also some misc Varnish changes that have been developed alongside this branch, but in Puppet. They need to be applied before this functionality will work.

Things still to do: 

- Provide a link from the index page to the R/P API index page. For testing, I was just going to /rpapi directly.
- If we enable Smart Bans in Varnish, then there's a TODO comment in controllers/rpapi.py to support Smart Bans from this app too

phg49389 committed:
Added first draft of RPAPI integration

phg49389 committed:
Enhanced Advanced Ban's validator

Now provides feedback to the user about what they’re doing wrong by
parsing the expression recursively and evaluating each argument and
phrase semantically, instead of a simple regular expression.

Also updated some bulk comments.

phg49389 committed:
Added some preliminary styling

Still has some fine-tuning to do, but it’s passable for now.

phg49389 committed:
Fleshed out templates with some actual content

phg49389 committed:
Updated Varnish backend domain list

phg49389 committed:
Improved Simple Ban behavior

Changed domain into a dropdown list and added a validator to the URL
field

phg49389 committed:
Refactored Thumbor into its own module, next to RPAPI

phg49389 commented:
Update: the Varnish-side changes have been implemented, so this will be ready to use as soon as it gets merged in.

phg49389 committed:
Some misc. cleanup before code review

phg49389 committed:
Fleshed out content of the index page and Thumbor's landing page

phg49389 committed:
Added h4 to RPAPI landing page to make it uniform

All the other cool pages have h4 tags!

phg49389 committed:
Adapted readmore() from Classifieds

Because no one likes walls of text

phg49389 committed:
Condensed the 3 subpages for each RPAPI action into the RPAPI landing page

phg49389 committed:
Removed unused example_script.js

Now that readmore.js is in the /static/js folder, it doesn’t need to
exist anymore.

phg49389 committed:
Styling tweaks

phg49389 committed:
Added links to the navbar

phg49389 committed:
Updated field names in views/rpapi.py to match the renames in forms/rpapi.py

phg49389 commented:
Thumbor has been moved into its own module, next to RPAPI. Navbar links to Thumbor and RPAPI have been added. Now there's only one thing left to do:

- If we enable Smart Bans in Varnish, then there's a TODO comment in controllers/rpapi.py to support Smart Bans from this app too

I think this is ready for a code review.

c-schwarze commented on app/__init__.py:
```diff
@@ -1,11 +1,18 @@
 
-cascade_connector = Cascade(app.config['SOAP_URL'], app.config['CASCADE_LOGIN'], app.config['SITE_ID'], app.config['STAGING_DESTINATION_ID'])
+# cascade_connector = Cascade(app.config['SOAP_URL'], app.config['CASCADE_LOGIN'],
+#                             app.config['SITE_ID'], app.config['STAGING_DESTINATION_ID'])
```
Do we not need this anymore? If we don't, we should delete it

phg49389 replied:
I think this was initially imported so we could use Cascade groups to determine who should be allowed to use this app, but then we migrated over to IAM groups via LDAP. I wasn't able to find any usages of `cascade_connector` in Cache Clear, so I just commented it out.

I agree though: if it's not used, it should be removed. I'll go do that.

c-schwarze reviewed (commented):
As stated in the github issue, we might be moving away from using varnish to cache images and things. If that is the case, I think what you are adding here will be removed. I'm not sure that we should add this in. We can chat more about this, if you would like

phg49389 committed:
Removed unused Cascade code

This has been obsolete since we changed over to IAM groups.

phg49389 commented:
I understand that we're not caching images in Varnish right now, but we're still caching static files like CSS, JS, and some font files, as well as certain routes in business-signage.bethel.edu and hall-of-fame.bethel.edu. Additionally, it's not unreasonable to anticipate that images could be cached once again. If I recall correctly, the reason we stopped caching images was due to triple-caching issues between Varnish and Thumbor. Thus, we only need to prevent cdn.bethel.edu/resize routes from being cached and the rest of images would be eligible for caching once more.

Setting aside the potential utility of this PR, this would complete the second half of the Refresh/Purge API that has been in Varnish for almost 11 months now. This PR has been planned for a long time, and I want to see it through even if it's not used right away. It is logical for me to write it since I authored the other half, and this prevents an awkward situation where someone else has to read through the VCL source code to understand how it works and read through issue 2 to see what I had in mind. A sort of "picking up where someone else left off" kind of situation, if you will.

c-schwarze commented:
👍 Alright, sounds good.

phg49389 committed:
Adapted FlaskView inheritance from MyBethel

This should prevent an issue where someone can go to /rpapi or /thumbor
without having their session set, which would through an error.
